domain/relation/service/relation.go

```diff
@@ -962,3 +987,120 @@ func (s *Service) SetRelationUnitSettings(
+			if err != nil {
+				return errors.Capture(err)
+			}
+		}
+	}
+	return nil
+}
+
+// ExportRelations returns all relation information to be exported for the
+// model.
+func (s *Service) ExportRelations(ctx context.Context) ([]relation.ExportRelation, error) {
+	relations, err := s.st.ExportRelations(ctx)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	// Generate the relation keys.
+	for i, r := range relations {
+		var eids []corerelation.EndpointIdentifier
+		for _, ep := range r.Endpoints {
+			eids = append(eids, corerelation.EndpointIdentifier{
+				ApplicationName: ep.ApplicationName,
+				EndpointName:    ep.Name,
+				Role:            ep.Role,
+			})
+		}
+		relations[i].Key, err = corerelation.NewKey(eids)
+		if err != nil {
+			return nil, errors.Errorf("generating relation key: %w", err)
+		}
+	}
+
+	return relations, nil
+}
+
+func (s *Service) importRelation(ctx context.Context, arg relation.ImportRelationArg) (corerelation.UUID, error) {
+	var relUUID corerelation.UUID
+
+	eps := arg.Key.EndpointIdentifiers()
+	s.logger.Criticalf(ctx, "importRelation(%d): %+v", arg.ID, eps)
+
+	switch len(eps) {
+	case 1:
+		// Peer relations are implicitly imported during migration of applications
+		// during the call to CreateApplication.
+		var err error
+		relUUID, err = s.st.GetPeerRelationUUIDByEndpointIdentifiers(ctx, eps[0])
+		if err != nil {
+			return relUUID, errors.Errorf("getting peer relation %d by endpoint %q: %w", arg.ID, eps[0], err)
+		}
+		// TODO - reset the relation ID.
+	case 2:
+		idep1, err := relation.NewCandidateEndpointIdentifier(eps[0].String())
+		if err != nil {
+			return relUUID, errors.Errorf("parsing endpoint identifier %q: %w", eps[0].String(), err)
+		}
+		idep2, err := relation.NewCandidateEndpointIdentifier(eps[1].String())
+		if err != nil {
+			return relUUID, errors.Errorf("parsing endpoint identifier %q: %w", eps[1].String(), err)
+		}
+
+		relUUID, err = s.st.AddRelationWithID(ctx, idep1, idep2, uint64(arg.ID))
+		if err != nil {
+			return relUUID, errors.Capture(err)
+		}
+	default:
+		return relUUID, errors.Errorf("unexpected number of endpoints %d for %q", len(eps), arg.Key)
+	}
+	return relUUID, nil
+}
+
+func (s *Service) importRelationEndpoint(ctx context.Context, relUUID corerelation.UUID, ep relation.ImportEndpoint) error {
+	appID, err := s.st.GetApplicationIDByName(ctx, ep.ApplicationName)
+	if err != nil {
+		return err
+	}
+
+	err = s.st.SetRelationApplicationSettings(ctx, relUUID, appID, settingsMap(ep.ApplicationSettings))
+	if err != nil {
+		return err
+	}
+	for unitName, settings := range ep.UnitSettings {
+		err = s.st.EnterScope(ctx, relUUID, unit.Name(unitName), settingsMap(settings))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func settingsMap(in map[string]interface{}) map[string]string {
+	return transform.Map(in, func(k string, v interface{}) (string, string) {
+		return k, fmt.Sprintf("%v", v)
```

Comment:
Don't do this.

Switch on the types you do expect and convert them to a string. There could be subtle differences between different go versions.

Reply:
This is from https://github.com/juju/juju/pull/19597 which this PR is rebased on.